aisynphys/pipeline/multipatch/cortical_location.py

```diff
@@ -0,0 +1,112 @@
+"""
+For generating DB table describing a cells location within cortex, 
+and adding layer-aligned distance info to the Pair table
+"""
+from ..pipeline_module import DatabasePipelineModule
+from .experiment import ExperimentPipelineModule
+from aisynphys import lims
+import logging
+import numpy as np
+from neuroanalysis.util.optional_import import optional_import
+get_depths_slice = optional_import('aisynphys.layer_depths', 'get_depths_slice')
+
+
+class CortexLocationPipelineModule(DatabasePipelineModule):
+    """Imports cell location data for each experiment
+    """
+    name = 'cortical_location'
+    dependencies = [ExperimentPipelineModule]
+    table_group = ['cortical_cell_location']
+    
+    @classmethod
+    def create_db_entries(cls, job, session):
+        db = job['database']
+        expt_id = job['job_id']
+
+        expt = db.experiment_from_ext_id(expt_id, session=session)
+        slice_entry = expt.slice
+
+        try:
+            images = lims.specimen_images(slice_entry.lims_specimen_name)
+            images = [image for image in images if image.get('treatment')=='DAPI']
+            assert len(images) > 0
+            image_series_id = images[0].get('image_series')
+            image_series_resolution = images[0].get('resolution')
+        except (AssertionError, ValueError):
+            logging.warning("No LIMS image series found for slice.")
+            return
```

Comment:
If `create_db_entries` fails to create entries in the database, it should raise an exception so that the failure is correctly logged in the database. Raising an exception here also causes the database session to be rolled back.

If `create_db_entries` succeeds in creating database entries, then just returning without exception will cause the session to be committed. 

Sometimes there are errors, but we create new database entries anyway. In that case you can return a list of error messages to be logged.

Reply:
Is the issue just that errors should be returned also, instead/in addition to using logging? Or are you saying I should create empty entries when there's no data?

Reply:
The logging is nice while running the pipeline, but later on we want to be able to ask why experiment 12345.678 doesn't have an entry in the cortical_cell_location table, or why it has an entry with null values, etc. The DB has a `pipeline` table for this purpose, where it keeps track of every job that was run, whether it succeeded, and whether the job returned any error messages. So all you need to do here is either (1) raise an exception, or (2) create the `cortical_cell_location` entry and return a message stating why some (or all) values in that entry are bad. 

The reasons you would pick one or the other depend on whether you think the issue is fixable. If you raise an exception here, then running an update on the database with `--retry` will cause that job to be attempted again in the future (which is helpful if a bug has been fixed, or the LIMS data has been filled in, etc.). On the other hand, if you create an entry along with an error message, that's more like saying "this job is complete; it has some problems, but they are permanent and it probably won't help to try running the job again".